### ISSUE TITLE
Performance issues when Tile is not found in Cache

### ISSUE THREAD

**Issue by JonasDaWi** [osmdroid/osmdroid#1124]
## Issue Type

[ ] Question
[x] Bug
[ ] Improvement
[ ] Build system related
[x] Performance
[ ] Documentation

## Description and/or steps/code to reproduce the problem
I have custom map tiles of a building from zoom level 15-23 created with mobac - osmdroid zip. They are displayed correctly. The problem is that when i rotate the map or update my position on the map using a locationprovider the performance gets really bad.
I profiled my application:
![photo_2018-08-02_10-04-26](https://user-images.githubusercontent.com/5794789/43635564-c1004874-970f-11e8-83b9-5f7b96d3c3ac.jpg)
osmdroid tries to load tiles that are not in my osmdroid tile zip. Also there are many errors ("SqlCache - Tile doesn't exist") when i rotate the map/update my position on the map.

My questions are:
1. How can I disable reloading tiles when i rotate the map or update my position on the map?
2. How can i prevent osmdroid to load non-existant tiles? 
3. Is it possible to disable the cache?

## Environment
Performance is really bad on older smartphones and tablets, like Samsung Galaxy S5 / Nexus 6. On newer smartphones the error also appears but does not lead to such a bad performance

### If it's a bug, version(s) of android this affects:

All

### Version of osmdroid the issue relates to:

6.0.2

**Comment by monsieurtanuki** [comment]
I don't think rotating or updating your position using a location provider has anything to do with it. 
Isn't it the same when you just pan the map? Maybe when you rotate you change the zoom level - that could explain why it happens when rotating. When you change the zoom level you have many new tiles to find, including the one used for the approximations.

I'm not familiar with the zip files in osmdroid: is that `MapTileAssetsProvider` that you use?

When osmdroid tries to find a tile, it uses its memory cache, its Sql cache and then its several providers.
By the way "SqlCache - Tile doesn't exist" is not an error, it's just a debug message. My first suggestion in order to improve performances: don't show the debug messages ;)

I assume the problem comes from your zip tile provider, that is too slow to say "I don't have this tile".
I think a solution would be - for tile providers that are slow to respond and with a pre-determined set of tiles - to add the concept of a set of tiles (and zoom levels) that don't exist with this provider. Therefore the provider will quickly say: don't bother me with this zoom level or with this tile, we both know I don't have it, you already asked me seconds before and there's no way it's going to improve.

@JonasDaWi What do you think of it?

**Comment by JonasDaWi** [comment]
It is the same when i pan the map. I think I don't change the zoom level when i rotate like this:
mMapView.setMapOrientation(bearing);
isn't it?
When I comment out this line, no tiles are loaded. So it should have something to do with it i think.
Also when I comment out the codepart where i change my position using a MyLocationNewOverlay using a LocationProvider new tiles are tried to load.
I think this should be improved within the library.


Regarding the tile provider I think your idea is worth trying.
I set the tile provider as following:
ITileSource tileSource = new XYTileSource("tilezip", 15, 22, 256, ".png", new String[]{""});
mMapView.getTileProvider().clearTileCache();
mMapView.getTileProvider().getTileWriter().onDetach();
mMapView.getTileProvider().detach();
mMapView.setTileProvider(new MapTileProviderBasic(getContext(), tileSource));

Is there something wrong with this?

**Comment by monsieurtanuki** [comment]
As stated previously, I'm not an expert on zip files for osmdroid. This is what I did:
* I created via MOBAC a zip file (format "Osmdroid ZIP")
* For some reason the tile style was "4uMaps"
* I put my file (named "4uMaps.zip") in my smartphone phone folder "osmdroid" (at the same level as "tiles")
* if I open my zip file, I only have one folder "4uMaps", then inside it one folder per zoom level and so on
* this is the code I used in order to use _only_ this file:
````java
final ITileSource source = new XYTileSource("4uMaps", 0, 19, 256, ".png", new String[] {""});
final MapTileFileArchiveProvider archiveProvider = new MapTileFileArchiveProvider(
    new SimpleRegisterReceiver(inflater.getContext()), source);
mMapView = new MapView(
    inflater.getContext(),
    new MapTileProviderArray(
        source,
        new SimpleRegisterReceiver(inflater.getContext()),
        new MapTileModuleProviderBase[]{archiveProvider}));
````
[Of course I strongly recommend you to match `MapTileProviderBasic` as much as possible, including `SqlTileWriter`, `MapTileApproximater` and so on, but this code is just for a test]

And the result is that I see my "4uMaps" tiles and only them. Which sounds like what you want, except that you encountered performance issues, right?

Regarding your questions about rotating, location provider, panning and cache, believe me: the problem is not within the cache or how often the tiles are requested, the problem lies with your provider.

I think size matters. My file is very small (6 tiles, 293 Kb): how large (number of tiles, file size) is yours?

**Comment by JonasDaWi** [comment]
I do almost exactly as you described.

My tile archive is 8.5MB, having 2748 Tiles. If I comment out the code for the rotation/updating my position on the map there is no cache error. 

What is the fastest tilesource for offline tiles? I have the .osm file of the offline area.

**Comment by monsieurtanuki** [comment]
I believe there are 2 problems:
1. you don't seem to use the `SqlTileWriter` cache
1. the time spent trying to get the tile from a zip, which is a problem if you don't have the tile (endlessly trying to get the tile from a hopeless and slow zip file) and if you have the tile (if you don't use the `SqlTileWriter` cache)

I'll try with such a big tile archive, in order to check if I'm right.
The solutions will be easy to implement. I'll tell you.

**Comment by JonasDaWi** [comment]
I wrote an [example application](https://github.com/JonasDaWi/OSMDroidSample) to observe the issues. In the following I describe the 3 Buttons and the issues they shall demonstrate:

### Toggle Rotation
When rotation is enabled the CPU usage goes up to about 20% on my Nexus 7. I rotate the map via  `mMapView.setMapOrientation()`. Is there a faster way to do a map rotation? Do other formats (other than .png file archives) rotate faster? 

### Toggle Position
When the position is enabled (NewPositionOverlay), the CPU usage rises by another 20%. Is that normal?

### Toggle Tileprovider
When i switch the TileProvider from your aforementioned `MapTileFileArchiveProvider` to `MapTileProviderBasic `the performance is horrible due to the already mentioned tile caching errors

> SqlCache - Tile doesn't exist
> Archives - Tile doesn't exist:

I think the `MapTileProviderBasic` should handle the caches better himself.
And the error I mentioned above: Stopping the rotation and only adding the position creates these caching errors, too. Is this a bug in the library?

One more question: What is the fastest tilesource?

**Comment by monsieurtanuki** [comment]
Quick answers:
* when rotation is activated the CPU usage goes up - it's normal as in that case we need to apply a rotation matrix to the canvas - cf. `Projection.save` - which is not the case in the more standard rotation==0 mode where it's optimized and we don't apply any matrix
* I couldn't find `NewPositionOverlay` in your code or in osmdroid's - is that in the bonus pack? If it's about GPS location, I think it's a battery consuming operation in general, which matches your CPU usage rise description
* again, `SqlCache - Tile doesn't exist` is not an error, it's a debug message. It means that the tile was not found in the SqlCache, which means it was never found before with some kind of `MapTileDownloader`, which sounds like a reasonable message if you don't use any `MapTileDownloader`.
* `Archives - Tile doesn't exist` that you mention here for the first time is deceiving - it means that the provider is going to search for the tile (in the zip file in your case)
* I don't know what the fastest tile format is. I guess SQLite is.

Solutions and questions for you:
1. You didn't answer my question - is your zip file in a phone folder or in an external SD card?
1. Are you OK with the performances with only the `MapTileFileArchiveProvider`? If not, how long does it take to get a tile in `MapTileFileArchiveProvider.TileLoader.loadTile`?
1. Could you also test the performances with the debug mode off?
1. If we want to understand more about your slow performances, I would suggest you to copy `MapTileProviderBasic` into `MyMapTileProviderBasic`, and then to run tests getting rid of this or that in order to understand what causes the slow performances. For instance, in your case I don't think you need the `assetsProvider` (your files are not embedded in the app), the `cacheProvider` (and you'll get rid of the `SqlCache` messages), the `approximationProvider` (as everything is allegedly embedded in your zip file) and the `downloaderProvider` (as you don't download, do you?). Not much left, is there? Perhaps the next thing to test/remove would be the "protected-cache-tile computers" and "pre-cache providers" lines at the end (but you need the "tiles currently being processed" at the very end

**Comment by JonasDaWi** [comment]
Sorry, i meant `MyLocationNewOverlay`, which is part of osmdroid. Please see my [demonstator project](https://github.com/JonasDaWi/OSMDroidSample), that demonstrates my issues.

Answers to you questions & observations:
1. It is a .zip file that is copied from the asset folder to the external sd card to the /osmdroid folder
2. loading a tile with the `MapTileFileArchiveProvider` takes between 10-35ms. This is acceptable, but not fast I'd say
3. It's the same with debug mode off
4. I did as you said, if I comment out `protected-cache-tile computers` in `MyMapTileProviderBasic` it is much faster! Please see also my demonstator app, where you comment out the `protected-cache-tile computers` under line 93. What does this mean?

**Comment by monsieurtanuki** [comment]
Quick answers:
1. Just curious: why do you have to copy it? Why don't you keep it in the asset folder?
1. OK
1. OK
1. There are several reasons why the pre-cache could be slow on your old smartphones. First, the pre-cache pre-computations are currently being optimized (by me). Second, the actual pre-cache process searchs for tiles in the zip file: if the access is slow on your device, or synchronous, then it's a problem for performances. Third, in the case of zip files we don't know anything _a priori_ about the file contents, which can be an issue when you have slow file access. A solution would be to say "by the way for this file I know I only have those zoom levels and those bounding boxes, so don't even try to get tile outside of that", but that would say extra parameters, less flexibility, and a small dev. Another solution would be to keep a `Set` of all the tiles tried once but not found - there's no need to check again those tiles in the zip file - but that means extra memory usage, and a small dev.

For the record the lines you commented are about 2 brother concepts:
* the _protected-cache-tiles_, where you say you want to keep the neighboring tiles (zoom+-1, screen border) in the memory cache, because you know you'll probably use them very soon
* the _pre-cache_, where you say that you'd like to even pre-cache the (same) neighboring tiles asynchronously in the memory cache, in order e.g. to avoid hiccups when you change zoom (waiting for the new zoom tile to be loaded into the memory cache)

Your best choice would be perhaps to keep the _protected-cache-tiles_, but to remove the _pre-cache-providers_.

**Comment by monsieurtanuki** [comment]
@JonasDaWi The code where I optimize the map tile cache access was merged about a week ago in #1131.
Could you try with this new version? That may solve your performance issues.

**Comment by monsieurtanuki** [comment]
@JonasDaWi ping
The fix is available in the new 6.0.3 release.
